### PR TITLE
KTOR-7675 Support CIO client for wasm-js and js

### PR DESCRIPTION
**Pull request by whyoleg**
**Subsystem**
Client CIO

**Motivation**
Fixes: https://youtrack.jetbrains.com/issue/KTOR-7675/Client-CIO-engine-support-for-wasm-js-and-js
It's a preparation for server CIO support for wasm-js and js.

**Solution**
Most of the code in `ktor-client-cio` is just copied with minor changes.
The most of the changes are related to two general parts in client tests:
* support for client engines registration for wasm-js and js too support automatic engine selection and allows to dynamically access registered engines in `ktor-client-tests`
* commonize `ktor-client-tests` code and align logic of filtering between platforms

### CODE REVIEW

File: ktor-client/ktor-client-core/jvm/src/io/ktor/client/HttpClientJvm.kt
Comment (whyoleg) [inline ktor-client/ktor-client-core/jvm/src/io/ktor/client/HttpClientJvm.kt:37]: Hey @osipxd!
I saw you created an issue (https://youtrack.jetbrains.com/issue/KTOR-7698) and wanted to tell, that I remembered doing it :)

and so... here are the changes. I've totally forgotten to extract them into a separate PR :(
I will then proceed with extracting those changes in separate PR.
Note: the pattern used here is implemented based on https://github.com/whyoleg/sweet-spi/pull/2 where I've already performed some extensive testing, that it works :)
Just telling it, so that you will not need to do this one more time :)

**Review by osipxd — COMMENTED** [review]
Thank you for the pull request!
I have just a few minor comments.

File: ktor-client/ktor-client-core/jsAndWasmShared/src/io/ktor/client/engine/Loader.kt
Comment (whyoleg) [reply ktor-client/ktor-client-core/jsAndWasmShared/src/io/ktor/client/engine/Loader.kt:18]: Yes! As well with a lot of other places where there are duplications like this.
I was started to think about it specifically when started working on adding support for wasm-wasi.
Here is the WIP sneak-peak of duplication removals: https://github.com/whyoleg/ktor/commit/5a4a5cfb5801573873f6e45543e805c85a088444

So, I can move `engines` to `nonJvm` now, or later, after all nodejs related PR's will be merged, create a PR for all changes at once - whatever will be better for you to review

File: ktor-client/ktor-client-core/jvm/src/io/ktor/client/HttpClientJvm.kt
Comment (whyoleg) [reply ktor-client/ktor-client-core/jvm/src/io/ktor/client/HttpClientJvm.kt:46]: I will drop changes in this file from this PR.

Yeah, we could drop `toList` here, but overall, it's more of a safeguard for not calling `ServiceLoader.load` multiple times: as `Iterable { ... }` factory will call lambda (creating `Iterator`), and so `ServiceLoader.load` each time it's accessed.
e.g:
```kotlin
val engines = Iterable { ... }

val a1 = engines.firstOrNull() // will create iterator
val a2 = engines.firstOrNull() // will create one more iterator
```

Just to keep in mind, and probably in this case would be nice to document this somewhere

File: ktor-client/ktor-client-core/jsAndWasmShared/src/io/ktor/client/engine/Loader.kt
Comment (whyoleg) [reply ktor-client/ktor-client-core/jsAndWasmShared/src/io/ktor/client/engine/Loader.kt]: nevermind, I just moved it to `nonJvm` in this PR :)

File: ktor-client/ktor-client-tests/common/src/io/ktor/client/tests/utils/ClientLoader.kt
Comment (osipxd) [reply ktor-client/ktor-client-tests/common/src/io/ktor/client/tests/utils/ClientLoader.kt:25]: [KTOR-7723](https://youtrack.jetbrains.com/issue/KTOR-7723) Engine exclusion from clientTests is confusing

**Review by osipxd — APPROVED** [review]
LGTM! Thanks

**Comment by whyoleg** [comment]
@osipxd, just curious, is there anything else needed from my side, so that the PR could be merged?
I'm already ready to create the last PR in a series with server support :)

**Comment by osipxd** [comment]
Nothing! I was just waiting for CI and then forgot to merge this.
Thank you for pinging me!

**Pull request merged**
